File: serverless-rest-api/javascript-http-sam/src/api/locations.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0

// Implementation of the API backend for locations

const { DynamoDB } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocument } = require("@aws-sdk/lib-dynamodb");
const { metricScope } = require("aws-embedded-metrics");
const uuid = require("uuid");
ddbClient = new DynamoDB({});

// Check if code is running in SAM Local environment and disable X-Ray in such case
if (!process.env.AWS_SAM_LOCAL) {
  const AWSXRay = require("aws-xray-sdk-core");
  ddbClient = AWSXRay.captureAWSv3Client(new DynamoDB({}));
}

const dynamo = DynamoDBDocument.from(ddbClient);
const tableName = process.env.LOCATIONS_TABLE;

exports.handler = metricScope(metrics => async (event, context) => {
  let body;
  let statusCode = 200;
  const headers = {
    "Content-Type": "application/json"
  };

  // Initialize putting common business metrics using EMF
  var metricPayload = {}
  metrics.putDimensions({ Service: "Locations" });
  metrics.putMetric("ProcessedLocations", 1, "Count");
  metrics.setProperty("requestId", event.requestContext.requestId);
  metrics.setProperty("routeKey", event.routeKey);

  try {
    switch (event.routeKey) {
      // Get all locations
      case "GET /locations":
        // add business metrics for the route
        metricPayload.operation = "GET";
        // get data from the database
        body = await dynamo.scan({ TableName: tableName });
        break;
      // Location CRUD operations
      case "GET /locations/{locationid}":
        // add business metrics for the route
        metricPayload.operation = "GET";
        metricPayload.locationid = event.pathParameters.locationid;
        // get data from the database
        body = await dynamo
          .get({
            TableName: tableName,
            Key: {
              locationid: event.pathParameters.locationid
            }
          });
        break;
      case "DELETE /locations/{locationid}":
        // add business metrics for the route
        metricPayload.operation = "DELETE";
        metricPayload.locationid = event.pathParameters.locationid;
        // delete item in the database
        await dynamo
          .delete({
            TableName: tableName,
            Key: {
              locationid: event.pathParameters.locationid
            }
          });
        body = {};
        break;
      case "PUT /locations":
        let requestJSON = JSON.parse(event.body);
        // generate unique id if it isn't present in the request
        let id = requestJSON.locationid || uuid.v1();
        let newitem = {
          locationid: id,
          timestamp: new Date().toISOString(),
          description: requestJSON.description,
          imageUrl: requestJSON.imageUrl,
          name: requestJSON.name
        }
        // add business metrics for the route
        metricPayload.operation = "PUT";
        metricPayload.locationid = id;
        // update the database
        await dynamo
          .put({
            TableName: tableName,
            Item: newitem
          });
        body = newitem;
        break;
      default:
        throw new Error(`Unsupported route: "${event.routeKey}"`);
    }
  } catch (err) {
    statusCode = 400;
    body = err.message;
    console.error(err.message);
  } finally {
    // Add route specific business metrics
    metrics.setProperty("Payload", metricPayload);
    body = JSON.stringify(body);
  }

  return {
    statusCode,
    body,
    headers
  };
});

File: python-rest-terraform/application/src/api/locations.py
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

# Implementation of the API backend for locations
import json
import uuid
import os
import boto3
from datetime import datetime
from aws_embedded_metrics import metric_scope

# Patch libraries to instrument downstream calls
# See https://docs.aws.amazon.com/xray/latest/devguide/xray-sdk-python-patching.html for more details
from aws_xray_sdk.core import patch_all
patch_all()

# Prepare DynamoDB client
LOCATIONS_TABLE = os.getenv('LOCATIONS_TABLE', None)
dynamodb = boto3.resource('dynamodb')
ddbTable = dynamodb.Table(LOCATIONS_TABLE)


@metric_scope
def lambda_handler(event, context, metrics):
    route_key = f"{event['httpMethod']} {event['resource']}"

    # Set default response, override with data from DynamoDB if any
    response_body = {'Message': 'Unsupported route'}
    status_code = 400
    headers = {
        'Content-Type': 'application/json',
        'Access-Control-Allow-Origin': '*'
        }

    # Initialize putting common business metrics using EMF
    metric_payload = {}
    metrics.put_dimensions({'Service': 'Locations'})
    metrics.put_metric('ProcessedLocations', 1, 'Count')
    metrics.set_property('requestId', event['requestContext']['requestId'])
    metrics.set_property('routeKey', route_key)

    try:
        # Get all locations
        if route_key == 'GET /locations':
            # generate business metrics for the route
            metric_payload['operation'] = 'GET'
            ddb_response = ddbTable.scan(Select='ALL_ATTRIBUTES')
            # return list of items instead of full DynamoDB response
            response_body = ddb_response['Items']
            status_code = 200
        # Location CRUD operations
        if route_key == 'GET /locations/{locationid}':
            # generate business metrics for the route
            metric_payload['operation'] = 'GET'
            metric_payload['locationid'] = event['pathParameters']['locationid']
            # get data from the database
            ddb_response = ddbTable.get_item(
                Key={'locationid': event['pathParameters']['locationid']}
            )
            # return list of items instead of full DynamoDB response
            if 'Item' in ddb_response:
                response_body = ddb_response['Item']
            else:
                response_body = {}
            status_code = 200
        if route_key == 'DELETE /locations/{locationid}':
            # generate business metrics for the route
            metric_payload['operation'] = 'DELETE'
            metric_payload['locationid'] = event['pathParameters']['locationid']
            # delete item in the database
            ddbTable.delete_item(
                Key={'locationid': event['pathParameters']['locationid']}
            )
            response_body = {}
            status_code = 200
        if route_key == 'PUT /locations':
            request_json = json.loads(event['body'])
            request_json['timestamp'] = datetime.now().strftime("%Y-%m-%dT%H:%M:%S.%f")
            # generate unique id if it isn't present in the request
            if 'locationid' not in request_json:
                request_json['locationid'] = str(uuid.uuid1())
            # generate business metrics for the route
            metric_payload['operation'] = 'PUT'
            metric_payload['locationid'] = request_json['locationid']
            # update the database
            ddbTable.put_item(
                Item=request_json
            )
            response_body = request_json
            status_code = 200
    except Exception as err:
        status_code = 400
        response_body = {'Error:': str(err)}
        print(str(err))
    # Add route specific business metrics
    metrics.set_property("Payload", metric_payload)
    return {
        'statusCode': status_code,
        'body': json.dumps(response_body),
        'headers': headers
    }
